superset/migrations/shared/catalogs.py

```diff
@@ -183,25 +199,16 @@ def downgrade_catalog_perms(engines: set[str] | None = None) -> None:
         if catalog is None:
             continue
 
-        # update schema_perms
-        ssh_tunnel = DatabaseDAO.get_ssh_tunnel(database.id)
-        for schema in database.get_all_schema_names(
-            catalog=catalog,
-            cache=False,
-            ssh_tunnel=ssh_tunnel,
-        ):
-            perm = security_manager.get_schema_perm(
+        try:
+            downgrade_schema_perms(database, catalog, session)
+        except Exception:  # pylint: disable=broad-except
```

Comment:
What happens if we can't downgrade but the app still downgrades? I'm assuming the user won't experience any errors since catalog_perms may exist but nothing is calling it.

Reply:
Hmm, good point.

The biggest problem with that is not the catalog permissions, but the schema permissions. For example, imagine we have a single table `catalog1.schema1.table1`, and a dataset `dataset1` on it. On an upgrade, the migration will:

1. Create the perm `catalog access on catalog1`
2. Rename the perm `schema access on schema1` to `schema access on catalog1.schema1`
3. Rename `dataset1.schema_perms` from `[database1].[schema1]` to `[database1].[catalog1].[schema1]`.

If we do a downgrade and it fails connecting to the analytical database, we would still have the new perms (1 and 2), but 3 would be reverted.

Now the question is, "would a user with access to `schema1` before the downgrade still be able to access `table1`"? And the answer is NO, because we do a string comparison to see if the dataset permissions match any of the user's DAR permissions.

So it seems to me that we need to change the downgrade to rename existing perms even if it can't connect to the analytical database. I'll change the logic so that if we can't get the list of schemas we update all the known schemas (by reading them from current perms, both on upgrade and downgrade). This is a much better solution.